src/hazelcore/semantics/CursorInfo.re

```diff
@@ -79,10 +79,18 @@ type typed =
   | OnRule
   | OnOp;
 
+[@deriving sexp]
+type uses_list = list(CursorPath.steps);
+
+[@deriving sexp]
+type pat_node =
+  | VarPat(UHPat.t, uses_list)
```

Comment:
The VarPat case should only occur when the UHPat.t is a variable pattern, not an arbitrary pattern. Following the principle of "make invalid states unrepresentable", let's change this to VarPat(ErrStatus.t, VarErrStatus.t, Var.t, uses_list). You can reconstruct the UHPat.t from the first three arguments when necessary.

Reply:
Since VarPat case only occur when UHPat.t is a variable pattern without Hole, can I just have `VarPat(Var.t, uses_list)`? BTW, UHPat.t can still be reconstructed by calling constructor of NotInHole and NotInVarHole when constructing UHPat.Var

Reply:
Since VarPat case only occur when UHPat.t is a variable pattern without Hole, can I just have `VarPat(Var.t, uses_list)`? Also, UHPat.t can still be reconstructed by calling constructor of NotInHole and NotInVarHole when constructing UHPat.Var

Reply:
ah yeah, good point!